cassandra/metadata.py

```diff
@@ -330,14 +331,25 @@ def add_or_return_host(self, host):
         """
         with self._hosts_lock:
             try:
-                return self._hosts[host.endpoint], False
+                return self._hosts[host.host_id], False
             except KeyError:
-                self._hosts[host.endpoint] = host
+                self._host_id_by_endpoint[host.endpoint] = host.host_id
+                self._hosts[host.host_id] = host
                 return host, True
 
     def remove_host(self, host):
         with self._hosts_lock:
-            return bool(self._hosts.pop(host.endpoint, False))
+            self._host_id_by_endpoint.pop(host.endpoint, False)
+            return bool(self._hosts.pop(host.host_id, False))
+
+    def remove_host_by_host_id(self, host_id):
+        with self._hosts_lock:
+            return bool(self._hosts.pop(host_id, False))
+
+    def update_host(self, host, endpoint):
+        host, created = self.add_or_return_host(host)
+        self._host_id_by_endpoint[host.endpoint] = host.host_id
```

Comment:
structure is modified without holding a lock

Reply:
i don't see any locking here. Did you forget to push some changes?

Reply:
I've stashed it by mistake, and forgot to put it back in